File: kotlin/src/main/kotlin/407.TrappingRainWaterII.kt
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.*

//https://leetcode.com/problems/trapping-rain-water-ii/


class TrappingRainWaterII{
  fun trapRainWater(heightMap: Array<IntArray>): Int {
    if(heightMap.isEmpty()) return 0
    val (n, m) = heightMap.size to heightMap[0].size
    val edges = PriorityQueue<Pair<Int, Int>>(compareBy{ heightMap[it.first][it.second]})
    val done = HashSet<Pair<Int, Int>>()
    for(i in 0 until n)
      for(j in 0 until m)
        if(i==0 || i==n-1 || j==0 || j==m-1) {
          edges.add(i to j)
          done.add(i to j)
        }

    var ret = 0
    while(edges.isNotEmpty()){
      val (i,j) = edges.poll()
      for((ni,nj) in listOf(i to j-1, i to j+1, i-1 to j, i+1 to j)){
        if(ni<0 || ni>=n || nj<0 || nj>=m || ni to nj in done)
          continue
        ret += maxOf(0, heightMap[i][j] - heightMap[ni][nj])
        heightMap[ni][nj] = maxOf(heightMap[ni][nj], heightMap[i][j])
        edges.add(ni to nj)
        done.add(ni to nj)
      }
    }
    return ret
  }

  @Test
  fun test1(){
    assertEquals(14, trapRainWater(arrayOf(
        intArrayOf(12,13,1,12),
        intArrayOf(13,4,13,12),
        intArrayOf(13,8,10,12),
        intArrayOf(12,13,12,12),
        intArrayOf(13,13,13,13)
    )))
  }
}

File: python/407_TrappingRainWaterII.py
from unittest import TestCase
# https://leetcode.com/problems/trapping-rain-water-ii/
import heapq


class TrappingRainWaterII(TestCase):
    def trapRainWater(self, heightMap):
        """
        :type heightMap: List[List[int]]
        :rtype: int
        """
        if not heightMap: return 0
        n, m = len(heightMap), len(heightMap[0])

        done = set()
        edges = []
        for i in range(n):
            for j in range(m):
                if i == 0 or i == n-1 or j == 0 or j == m-1:
                    done.add((i, j))
                    heapq.heappush(edges, (heightMap[i][j], i, j))

        ret = 0
        while edges:
            cur, i, j = heapq.heappop(edges)
            for (ni, nj) in ((i, j-1), (i, j+1), (i-1, j), (i+1, j)):
                if not (0 <= ni < n and 0 <= nj < m and (ni, nj) not in done):
                    continue
                h = max(cur, heightMap[ni][nj])
                ret += h - heightMap[ni][nj]
                done.add((ni, nj))
                heapq.heappush(edges, (h, ni, nj))
        return ret

    def test1(self):
        self.assertEqual(4, self.trapRainWater([
            [1,4,3,1,3,2],
            [3,2,1,3,2,4],
            [2,3,3,2,3,1]
        ]))

    def test2(self):
        self.assertEqual(14, self.trapRainWater([
            [12,13,1,12],
            [13,4,13,12],
            [13,8,10,12],
            [12,13,12,12],
            [13,13,13,13]
        ]))